### PR TITLE
Remove agressive precondition in zero-length-writes.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

The zero-length writes handler will crash if it receives channelInactive
before channelActive. Sadly, that's totally possible if a channel is
closed from the connect promise.

Modifications:

- Tone back the preconditions a bit.
- Add a test that triggers this path.

Result:

Channel handler behaves better on early close

### CODE REVIEW

**Review by glbrntt — APPROVED** [review]
This LGTM. 

For my own understanding: do we only expect `channelInactive` before `channelActive` if closing locally?

File: Sources/NIOTransportServices/NIOFilterEmptyWritesHandler.swift [hunk 1]
```diff
@@ -88,9 +88,16 @@ public final class NIOFilterEmptyWritesHandler: ChannelDuplexHandler {
...
-        context.fireChannelActive()
+        switch self.state {
+        case .notActiveYet:
+            self.state = .open
+            context.fireChannelActive()
+        case .closedFromLocal:
+            // Closed before we activated, not a problem.
+            assert(self.prefixEmptyWritePromise == nil)
+        case .open, .closedFromRemote, .error:
+            preconditionFailure()
```
Comment (Davidde94) [inline Sources/NIOTransportServices/NIOFilterEmptyWritesHandler.swift:99]: Not worth having a tiny little message here?

**Review by Davidde94 — APPROVED** [review]

File: Sources/NIOTransportServices/NIOFilterEmptyWritesHandler.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIOTransportServices/NIOFilterEmptyWritesHandler.swift:99]: I think it's largely unnecessary: on iOS crash reports will point directly at this line anyway, and the static string raises binary sizes with minimal improvement in debuggability.

**Comment by Lukasa** [comment]
> For my own understanding: do we only expect `channelInactive` before `channelActive` if closing locally?

Yes.

**Pull request merged**